packages/sdk/app-framework/src/plugins/plugin-intent/intent-dispatcher.ts

```diff
@@ -0,0 +1,281 @@
+ * @param params.filter Optional filter to determine if the resolver should be used.
+ */
+export const createResolver = <Tag extends string, Fields extends IntentParams>(
+  schema: IntentSchema<Tag, Fields>,
+  effect: IntentEffectDefinition<Fields>,
+  params: Pick<IntentResolver<Tag, Fields>, 'disposition' | 'filter'> = {},
+): IntentResolver<Tag, Fields> => ({
+  action: schema._tag,
+  effect,
+  ...params,
+});
+
+/**
+ * Invokes intents and returns the result.
+ */
+export type PromiseIntentDispatcher = <Fields extends IntentParams>(
+  intent: IntentChain<any, any, any, Fields>,
+) => Promise<IntentDispatcherResult<Fields>>;
+
+/**
+ * Creates an effect for intents.
+ */
+export type IntentDispatcher = <Fields extends IntentParams>(
+  intent: IntentChain<any, any, any, Fields>,
+  depth?: number,
+) => Effect.Effect<IntentDispatcherResult<Fields>, Error>;
+
+type IntentResult<Tag extends string, Fields extends IntentParams> = IntentEffectResult<Fields> & {
+  _intent: Intent<Tag, Fields>;
+};
+
+type AnyIntentResult = IntentResult<any, any>;
+
+/**
+ * Invokes the most recent undoable intent with undo flags.
+ */
+export type PromiseIntentUndo = () => Promise<IntentDispatcherResult<any> | undefined>;
+
+/**
+ * Creates an effect which undoes the last intent.
+ */
+export type IntentUndo = () => Effect.Effect<IntentDispatcherResult<any> | undefined, Error>;
+
+/**
+ * Check if a chain of results is undoable.
+ */
+const isUndoable = (historyEntry: AnyIntentResult[]): boolean =>
+  historyEntry.length > 0 && historyEntry.every(({ undoable }) => !!undoable);
+
+export type IntentContext = {
+  dispatch: IntentDispatcher;
+  dispatchPromise: PromiseIntentDispatcher;
+  undo: IntentUndo;
+  undoPromise: PromiseIntentUndo;
+  registerResolver: (id: string, resolver: AnyIntentResolver) => () => void;
+};
+
+/**
+ * Sets of an intent dispatcher.
+ *
+ * @param resolvers An array of available intent resolvers.
+ * @param params.historyLimit The maximum number of intent results to keep in history.
+ * @param params.executionLimit The maximum recursion depth of intent chains.
+ */
+export const createDispatcher = (
+  resolvers: Record<string, AnyIntentResolver[]>,
+  { executionLimit = EXECUTION_LIMIT, historyLimit = HISTORY_LIMIT } = {},
+): IntentContext => {
+  const historyRef = Effect.runSync(Ref.make<AnyIntentResult[][]>([]));
+
+  const handleIntent = (intent: AnyIntent) => {
+    return Effect.gen(function* () {
+      const candidates = Object.entries(resolvers)
+        .filter(([id, _]) => (intent.plugin ? id === intent.plugin : true))
+        .flatMap(([_, resolvers]) => resolvers)
+        .filter((r) => r.action === intent.action)
+        .filter((r) => !r.filter || r.filter(intent.data))
+        .toSorted(({ disposition: a = 'default' }, { disposition: b = 'default' }) => {
+          return a === b ? 0 : a === 'hoist' || b === 'fallback' ? -1 : b === 'hoist' || a === 'fallback' ? 1 : 0;
+        });
```

Comment:
I think it's less useful here than in the surfaces, but basically it allows you to define a conditional handler which you prioritize. The dynamic thread handlers I have implemented as hoisted to ensure they get placed first since dynamic handlers are no longer handled as a special case